packages/SwingSet/test/test-state.js

```diff
@@ -71,187 +70,67 @@ async function testStorage(t, s, getState, commit) {
 }
 
 test('storageInMemory', async t => {
-  const store = initSwingStore(null);
-  await testStorage(t, store.kvStore, () => getAllState(store).kvStuff, null);
-});
-
-test('crankBuffer fulfills storage API', async t => {
-  const store = initSwingStore(null);
-  const { crankBuffer, commitCrank } = buildCrankBuffer(
-    store.kvStore,
-    createSHA256,
-  );
+  const kernelStorage = initSwingStore(null).kernelStorage;
   await testStorage(
     t,
-    crankBuffer,
-    () => getAllState(store).kvStuff,
-    commitCrank,
+    kernelStorage.kvStore,
+    () => getAllState(kernelStorage).kvStuff,
+    null,
   );
 });
 
-test('crankBuffer handles key iteration properly even with intra-crank data changes', t => {
```

Comment:
I see why this goes away (the crankBuffer is not really buffering things anymore), but I need to think about whether the `getKeys` swingstore is now providing will do the right thing w.r.t. commit points. 95% confident that it does, but XXX-@warner remind myself to double check.

Reply:
Ok, yes, it should work just as well as it did before. Previously, the LMDB `getKeys()` gave us an [actual iterator](https://www.npmjs.com/package/lmdb#user-content-dbgetkeysoptions-rangeoptions-iterableany), anchored in a read transaction (so subsequent writes wouldn't change the data), swing-store exposed that directly (using `yield* ensureTxn().getKeys(rangeOptions)`), and `kernelSyscall.js > vatstoreGetAfter` used [`clearVatStoreIteration()`](https://github.com/Agoric/agoric-sdk/blob/4143288b0a030309c94c7dc2e8858318f26dea48/packages/SwingSet/src/kernel/kernelSyscall.js#L54) to abandon the iterator if it saw any sign of those writes, or if a different vat shows up to iterate (since LMDB only offered one iterator at a time, I think).

This new scheme uses a SQL query to fetch *all* rows (no iterator) with `keys = sqlKVGetKeys.all(..)`, then `yield* keys.values()` to return an iterator over that saved-in-RAM copy. `kernelSyscall.js` does the same as before, so it's abandoning that iterator under the same conditions. A vat might manage to hold onto the iterator for quite a while: in the LMDB scheme, that would have kept fetching data from the same read txn (and since nothing outside `kernelSyscall.js` would modify that portion of the data, the vatstore, this wouldn't cause any problems). In the SQLite scheme, the read txn is finished promptly, because we pull all values into RAM right away.

So the new scheme will use more RAM when getKeys() is used. The use cases I can find are:
* `vatstoreGetAfter` (called from vat userspace, one key at a time)
* `cleanupAfterTerminatedVat` (in kernelKeeper): to delete/orphan/reject c-list entries, during vat termination
* `getStaticVats` / `getDevices`: at startup to prepare all vats/devices
* `dumpState` (vatKeeper): to walk whole c-list
* `makeBufferedStorage` (lib/storageAPI.js): now unused by swingset, #6704 is to remove this
* a commented-out version of `getImporters` in src/vats/comms/state.js
  * the enabled version stores all the references in a single key, as a serialized list
  * we should decide whether it's worth doing something more sophisticated: the enabled version would be slow for lrefs that are shared among lots of importers, but I don't think that's generally the case for us

The `vatstoreGetAfter` could be large: virtual collections used this for any iterators (`.keys()`, `.values()`, `.entries()`). The swing-store will load the entire virtual collection into RAM, then feed out one entry at a time to `kernelSyscall.js`. We need to replace this with a `getNext()` -like one-key-at-a-time scheme before this gets released.

Also, the c-list deletion code in `cleanupAfterTerminatedVat` could be large, it will load a significant portion of the c-list into RAM at a time (first all the exported objects, then all the imported objects, then the promises). We can start by replacing this with `getNext()`, but hopefully we can replace it with a more dedicated "delete everything" API as part of #6677.

Reply:
> This new scheme uses a SQL query to fetch _all_ rows (no iterator) with `keys = sqlKVGetKeys.all(..)`, then `yield* keys.values()` to return an iterator over that saved-in-RAM copy.

> The `vatstoreGetAfter` could be large: virtual collections used this for any iterators (`.keys()`, `.values()`, `.entries()`). The swing-store will load the entire virtual collection into RAM, then feed out one entry at a time to `kernelSyscall.js`. We need to replace this with a `getNext()` -like one-key-at-a-time scheme before this gets released.

This concerns me. If there are millions of entries in a collection, we should not ever be holding all keys in RAM.

Furthermore, can a vat driven iteration cross crank boundaries? What happens to that in RAM state then?

I see 3 solutions:
- Change the iteration model so that insertions (and deletes) to the collection impact an existing iteration if the iterator has not yet reached that position. JS built-in collections will include newly inserted elements, but that's not fully relevant since they are iterated by in insertion order.
  - There ought to be an SQL query to select a single row after another one given some filter & ordering constraint?
- If we want to skip newly inserted entries (and I assume keep skipping deleted but not yet iterated on entries), SQLite in WAL mode does support read transactions like LMDB does. The problem is restoring these transactions across restarts. This is only a problem if the iteration needs to cross crank boundaries (well block boundaries really).
  - An alternative would be to have a table of "existing iterators" which references rows the iterator needs to iterate over, aka copy keys into a separate iteration table.

For context, at some point I knew how the different iterations worked, but I've completely forgotten, so I may be completely off the mark here.

Reply:
> can a vat driven iteration cross crank boundaries?

No.

Dealing with long lived/large iterators has been an issue in our queue for a while now.

Reply:
Which is to say, it's a big issue but not one for this PR.

Reply:
@mhofman writes:
> Furthermore, can a vat driven iteration cross crank boundaries? What happens to that in RAM state then?

There are three layers:
* a vat, where liveslots and `collectionManager.js` [create a user-facing iterator](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/liveslots/collectionManager.js#L433-L438) around a sequence of `vatstoreGetAfter()` syscalls, which take a couple of keys (prior, start, end) and maybe return a single key/value pair
* `kernelSyscall.js`, which [implements `vatstoreGetAfter()`](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/kernel/kernelSyscall.js#L162) around a [sometimes-long-lasting iterator](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/kernel/kernelSyscall.js#L51) that it [gets back from `kvStore.getKeys()`](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/kernel/kernelSyscall.js#L209)
* swingstore's implementation of `kvStore.getKeys()`, which [used to delegate to an LMDB iterator](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/swing-store/src/swingStore.js#L238-L254), and is [now implemented by iterating through a big Array](https://github.com/Agoric/agoric-sdk/blob/3087-sql-swingstore/packages/swing-store/src/swingStore.js#L305-L317) that it gets by immediately and completely consuming a short-lived SQLite query iterator

The `workingKeyIterator` in `kernelSyscall.js` can outlive a single crank, but it gets invalidated upon any vatstoreSet or vatstoreDelete, or if vatstoreGetAfter is called with a `priorKey` that doesn't match the remembered value. The actual `priorKey` comparison uses the output of `vatstoreKeyKey()`, which incorporates the vatID, which prevents different vats from observing each other's iterations. (I think I remember deciding that this was sound when we wrote this code, but now it's making me anxious and I'm looking forward to removing it).

> There ought to be an SQL query to select a single row after another one given some filter & ordering constraint?

Yeah, that's what the new `getNext` API will use: it will return the key after the one you asked about, perhaps subject to some other limits (like an upper bound), and can return `undefined` if there is no such key (within the range that the particular vat is allowed to read). The plan (well, my plan, I still need to discuss it with @FUDCo) is the bottom two layers to strictly use a stateless, non-iterating API, and if anything in the liveslots layer wants to provide an iterator, it can build one itself out of a series of `getNext` calls. At that point, the question of whether iterators are killed by insertions, or see only newer keys, or whatever, is up to liveslots, and we can figure out the tradeoffs (between "does it work just like JS native iterators" and "how hard is it to implement" and performance) at that layer.

Reply:
> > can a vat driven iteration cross crank boundaries?
> 
> No.



> > Furthermore, can a vat driven iteration cross crank boundaries? What happens to that in RAM state then?
> 
> There are three layers:
> 
> * a vat, where liveslots and `collectionManager.js` [create a user-facing iterator](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/liveslots/collectionManager.js#L433-L438) around a sequence of `vatstoreGetAfter()` syscalls, which take a couple of keys (prior, start, end) and maybe return a single key/value pair
> * `kernelSyscall.js`, which [implements `vatstoreGetAfter()`](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/kernel/kernelSyscall.js#L162) around a [sometimes-long-lasting iterator](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/kernel/kernelSyscall.js#L51) that it [gets back from `kvStore.getKeys()`](https://github.com/Agoric/agoric-sdk/blob/02e058d619f363341398c4f345010599d6b85166/packages/SwingSet/src/kernel/kernelSyscall.js#L209)

Sorry to keep digging here, but what prevents a vat from getting an iterator on a collection, keep it in its heap, and call `next()` on it across crank boundaries? Does liveslot actively drop any alive iterator at the end of a crank?

I only see the vat side iterator being invalidated on collection insertion, and the kernel side iterator being invalidated on vatstore operations. I don't see any mechanism where these iterator are invalidated across crank boundaries.

The risk of course is that the vat iterator would be relying on a DB read transaction / in memory structure that goes away on kernel restart.